tensorboard/uploader/exporter.py

```diff
@@ -180,19 +227,120 @@ def _request_scalar_data(self, experiment_id, read_time):
+            scalar data points.
+        """
+        wall_times = [t.ToNanoseconds() / 1e9 for t in points.wall_times]
+        return {
+            u"steps": list(points.steps),
+            u"wall_times": wall_times,
+            u"values": list(points.values),
+        }
+
+    def _process_blob_sequence_points(self, blob_sequences, experiment_id):
+        """Process blob sequence points.
+
+        As a side effect, also downloads the binary contents of the blobs
+        to respective files. The paths to the files relative to the
+        experiment directory is encapsulated in the returned JSON object.
+
+        Args:
+          blob_sequences:
+            `export_service_pb2.StreamDataResponse.BlobSequencePoints` proto.
+
+        Returns:
+          A JSON-serializable `dict` for the steps and wall_times, as well as
+            the blob_file_paths, which are the relative paths to the downloaded
+            blob contents.
+        """
+        wall_times = [
+            t.ToNanoseconds() / 1e9 for t in blob_sequences.wall_times
+        ]
+        json_object = {
+            u"steps": list(blob_sequences.steps),
+            u"wall_times": wall_times,
+            u"blob_file_paths": [],
+        }
+        blob_file_paths = json_object[u"blob_file_paths"]
+        for blobseq in blob_sequences.values:
+            seq_blob_file_paths = []
+            for entry in blobseq.entries:
+                if (
+                    entry.HasField("blob")
+                    and entry.blob.state
+                    == blob_pb2.BlobState.BLOB_STATE_CURRENT
+                ):
+                    blob_path = self._download_blob(
+                        entry.blob.blob_id, experiment_id
+                    )
+                    seq_blob_file_paths.append(blob_path)
+                else:
+                    seq_blob_file_paths.append(None)
+            blob_file_paths.append(seq_blob_file_paths)
+        return json_object
+
+    def _download_blob(self, blob_id, experiment_id):
+        """Download the blob via rpc.
+
+        Args:
+          blob_id: Id of the blob.
+          experiment_id: Id of the experiment that the blob belongs to.
+
+        Returns:
+          If the blob is downloaded successfully:
+            The path of the downloaded blob file relative to the experiment
+            directory.
+          Else:
+            `None`.
+        """
+        # TODO(cais): Deduplicate with internal method perhaps.
+        experiment_dir = _experiment_directory(self._outdir, experiment_id)
+        blobs_dir = os.path.join(experiment_dir, _BLOBS_DIR)
+        if not os.path.isdir(blobs_dir):
+            _mkdir_p(blobs_dir)
+        request = export_service_pb2.StreamBlobDataRequest(blob_id=blob_id)
+        blob_abspath = os.path.join(
+            blobs_dir, _BLOB_FILE_PREFIX + blob_id + _BLOB_FILE_SUFFIX
+        )
+        with open(blob_abspath, "wb") as f:
```

Comment:
Can this be `_open_excl(blobs_abspath, "wb")` (adding a `mode` argument
to `_open_excl` is fine), or is there a reason that we need to be able
to silently stomp this file?